### DISCUSSION TITLE
Now Available ✨ Use project status updates with GraphQL and webhooks

### DISCUSSION THREAD

**Discussion by github-staff** [community/community#129854] [Projects and Issues]
Today's **changelog** brings you GraphQL and webhook support for [project status updates](https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/sharing-project-updates) and project custom field changes directly in the webhook event!

## **⚙️ Using GraphQL and webhooks with project status updates**

Following our release [earlier this year for project status updates](https://github.blog/changelog/2024-01-18-github-issues-projects-project-status-updates-issues-side-panel/), you can now interact with project status updates using GraphQL and webhooks. This unlocks new ways to automate how you provide and gather project status update information.

### **GraphQL**

There is a new <code>[ProjectV2StatusUpdate](https://docs.github.com/en/graphql/reference/objects#projectv2statusupdate)</code> GraphQL object to interact with project status updates, so you can view, create, update, and delete status updates. 

Below is an example query to [create](https://docs.github.com/en/graphql/reference/mutations#createprojectv2statusupdate) a new project status update.

```
mutation {
  createProjectV2StatusUpdate(
    input: {projectId: "0123456", body: "We wrapped up our bug bash following the beta rollout. We're back on track for our GA date in August! 🚀", startDate: "2024-06-03", targetDate: "2024-08-09", status: ON_TRACK}
  ) {
    statusUpdate {
      id
      startDate
      targetDate
      body
      bodyHTML
      status
    }
  }
}

```

### **Webhooks**

Project status updates are included in the new <code>[projects_v2_status_update](https://docs.github.com/en/webhooks/webhook-events-and-payloads#projects_v2_status_update)</code> webhook event, so you can understand and be notified when a new project status update is provided.

You must be subscribed to this event from the organization settings page to receive this information. 

![343190942-6b58d44a-83fe-413d-8920-9a4ca166317c](https://github.com/community/community/assets/126821655/3272bdbb-2e7d-477b-bbb1-5f6f12041709)

Below is an example of a webhook event.

```
{
    "action": "edited",
    "projects_v2_status_update": {
        "id": 32633,
        "node_id": "PVTSU_lADOBH2n9s4Ajp6VzX95",
        "project_node_id": "PVT_kwDOBH2n9s4Ajp6V",
        "creator": {
          ...
        },
        "body": "We've kicked off this project and are feeling confident in our rollout plan. More updates and demos to come next week!",
        "start_date": "2024-06-24",
        "target_date": "2024-08-16",
        "status": "ON_TRACK",
        "created_at": "2024-06-24T20:27:48Z",
        "updated_at": "2024-06-24T20:30:47Z"
    },
    "changes": {
        "body": {
            "from": "We're still planning this out and are kicking off soon.",
            "to": "We've kicked off this project and are feeling confident in our rollout plan. More updates and demos to come next week!"
        },
        "status": {
            "from": "INACTIVE",
            "to": "ON_TRACK"
        },
        "start_date": {
            "from": null,
            "to": "2024-06-24"
        },
        "target_date": {
            "from": null,
            "to": "2024-08-16"
        }
    },
    "organization": {
        ...
    },
    "sender": {
        ...
    }
}

```

## **➕ Using webhooks for project custom field changes**

Project custom field changes are now included directly in the [`project_v2_item`](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=edited#projects_v2_item) webhook event when a project item's fields are edited, removing the need to send an additional GraphQL query. This gives you the previous and current field values to understand how project fields change over time and how long they have a particular value, allowing you to understand how long an item was `In progress` before moving to `Done` status. 

Below is an example of the webhook which includes the previous and current value for `single select`, `text`, `number`, `iteration`, and `date` project custom fields using the `changes` parameter.

```
"changes": {
    "field_value": {
        "field_node_id": "PVTSSF_lADOBH2n9s4Aje1Izgb1kEs",
        "field_type": "single_select",
        "field_name": "Status",
        "project_number": 18,
        "from": {
            "id": "f75ad846",
            "name": "Todo",
            "color": "GREEN",
            "description": "This item hasn't been started"
        },
        "to": {
            "id": "47fc9ee4",
            "name": "In Progress",
            "color": "YELLOW",
            "description": "This is actively being worked on"
        }
    }
},

```

## **✨ Bug fixes and improvements**

* Added the <code>[convertProjectV2DraftIssueItemToIssue](https://docs.github.com/en/graphql/reference/mutations#convertprojectv2draftissueitemtoissue)</code> GraphQL mutation to convert drafts to issues
* Fixed an error message when resizing columns in the table layout
* Fixed errors when migrating a classic project to the new Projects experience
* Fixed a bug where updating an issue in the project side panel didn't reflect in the project view
* Fixed the rendering of special characters in a single-select field description from the table layout cell dropdown
* Fixed a bug where a space could not be added in project chart titles

## <strong>✍️ Tell us what you think!</strong>

Share your feedback by posting a comment below 👇 How are you going to use this in your projects?

See how to use GitHub for project planning with [GitHub Issues](http://github.com/features/issues), check out what's on the [roadmap](https://github.com/orgs/github/projects/4247/views/7), and learn more in the [documentation](https://docs.github.com/issues).

**Comment by zsloane** [comment]
Is this available in Actions or only for webhooks to external services?

**Reply by rileybroughten** [reply]
This is only available for webhooks to external services, as action workflows are currently only triggered by events in your repository, rather than projects that live at the organization level.

**Comment by Aerex** [comment]
Sorry for double positing. I made a comment in a related **[issue](https://github.com/orgs/community/discussions/17405#discussioncomment-12029887)**. This feature does not appear to work for me. I have subscribed to the event on my GitHub App project and I don't get any events delivered. I have check permissions and I am subscribed to the correct events

![image](https://github.com/user-attachments/assets/4c9eaa86-aaf6-4c19-af5c-5c5dd65acb01)

![image](https://github.com/user-attachments/assets/df874bc5-f4b3-4c11-962f-cb16d075da12)
